3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,202 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {IframeMessagingClient} from './iframe-messaging-client';
+import {map} from '../src/utils/object';
+import {dev} from '../src/log';
+import {initLogConstructor, setReportError} from '../src/log';
+import {reportError} from '../src/error';
+initLogConstructor();
+setReportError(reportError);
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * This provides the "glue" between the AMP Analytics tag and the third-party
+ * vendor's metrics-collection page.
+ *
+ * The vendor's page must implement onNewAmpAnalyticsInstance().
+ */
+class AmpAnalytics3pRemoteFrameHelper {
+  constructor(win) {
+    /**
+     * @type {!Array<function(string)>}
+     * @private
+     */
+    this.eventsListeners_ = [];
+
+    /**
+     * @type {!Array<function(string,string)>}
+     * @private
+     */
+    this.extraDataListeners_ = [];
+
+    /**
+     * @type {!Array<Object>}
+     * @private
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * @type {!Array<Object>}
+     * @private
+     */
+    this.receivedExtraDataQueue_ = [];
+
+    /**
+     * Holds a mapping between sender ID and extra data
+     * @type {Map}
+     * @private
+     */
+    this.senderIdToExtraData_ = map();
+
+    /**
+     * Simple requestIdleCallback polyfill
+     * @type {Function<Function>}
+     * @private
+     */
+    this.requestIdleCallback_ =
+      win.requestIdleCallback.bind(win) || (cb => setTimeout(cb, 1));
+
+    /**
+     * Handles communication between frames
+     * @type {IframeMessagingClient}
+     * @private
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(JSON.parse(window.name).sentinel);
+    this.iframeMessagingClient_.registerCallback('ampAnalytics3pMessages',
+      this.onReceivedMessagesFromCreative_.bind(this));
+
+    this.sendReadyMessageToCreative_();
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur
+   * @param {!Function} listener A function that takes an array of event
+   * strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    this.eventsListeners_.push(listener);
+  }
+
+  /**
+   * Registers a callback function to be called when extra setup data is
+   * recevied from an AMP Analytics tag.
+   * @param {!Function} listener A function to receive any extra
+   * setup data sent by the creative
+   */
+  registerAmpAnalytics3pExtraDataListener(listener) {
+    this.extraDataListeners_.push(listener);
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @param {!string} message The message to send.
+   */
+  sendMessageToCreative(message) {
+    this.iframeMessagingClient_.sendMessage('ampAnalytics3pResponse',
```

Comment:
Consider making message types an enum and you could then have one central messaging function.   Something like:

/** @private @const @enum {string} */
const MESSAGE_TYPE { READY: 'ampAnalytics3pReady', RESPONSE: 'ampAnalytics3pResponse' };

/**
 * @param {!MESSAGE_TYPE} type
 * @param {Object<string,*>=} opt_content
 * @private
 */
sendMessageToCreative_(type, opt_content) {
   const content = opt_content || {};
   content['senderId'] = this.id_;
   this.iframeMessageClient_.sendMessage(type, content);
}